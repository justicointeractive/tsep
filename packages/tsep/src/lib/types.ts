export type Primitive = string | number | boolean | RegExp | null | undefined;

export interface Expression {
  type: string;
  [key: string]: Primitive | Expression | Array<Primitive | Expression>;
}

export interface ArrayExpression extends Expression {
  type: 'ArrayExpression';
  elements: (Expression | undefined)[];
}

export interface BinaryExpression extends Expression {
  type: 'BinaryExpression';
  operator: string;
  left: Expression;
  right: Expression;
}

export interface CallExpression extends Expression {
  type: 'CallExpression';
  arguments: Expression[];
  callee: Expression;
}

export interface Compound extends Expression {
  type: 'Compound';
  body: Expression[];
}

export interface ConditionalExpression extends Expression {
  type: 'ConditionalExpression';
  test: Expression;
  consequent: Expression;
  alternate: Expression;
}

export interface Identifier extends Expression {
  type: 'Identifier';
  name: string;
}

export interface Literal extends Expression {
  type: 'Literal';
  value: boolean | number | string | RegExp | null;
  raw: string;
}

export interface MemberExpression extends Expression {
  type: 'MemberExpression';
  computed: boolean;
  object: Expression;
  property: Expression;
  optional?: boolean;
}

export interface ThisExpression extends Expression {
  type: 'ThisExpression';
}

export interface UnaryExpression extends Expression {
  type: 'UnaryExpression';
  operator: string;
  argument: Expression;
  prefix: boolean;
}

export type ExpressionType =
  | 'Compound'
  | 'Identifier'
  | 'MemberExpression'
  | 'Literal'
  | 'ThisExpression'
  | 'CallExpression'
  | 'UnaryExpression'
  | 'BinaryExpression'
  | 'ConditionalExpression'
  | 'ArrayExpression';

export type CoreExpression =
  | ArrayExpression
  | BinaryExpression
  | CallExpression
  | Compound
  | ConditionalExpression
  | Identifier
  | Literal
  | MemberExpression
  | ThisExpression
  | UnaryExpression;

export type BinaryOpInfo = { value: string; prec: number; right_a: boolean };
