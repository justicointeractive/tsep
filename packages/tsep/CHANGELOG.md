# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

### [0.0.1-alpha.2](https://github.com/justicointeractive/tsep/compare/tsep-0.0.1-alpha.1...tsep-0.0.1-alpha.2) (2022-08-01)

### [0.0.1-alpha.1](https://github.com/justicointeractive/tsep/compare/tsep-0.0.1-alpha.0...tsep-0.0.1-alpha.1) (2022-08-01)


### Bug Fixes

* export types ([47a8092](https://github.com/justicointeractive/tsep/commit/47a809273ca2fdbe935e19d429751656e102833c))

### 0.0.1-alpha.0 (2022-08-01)
