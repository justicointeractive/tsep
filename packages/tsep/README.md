# tsep

tsep is a fork of https://www.npmjs.com/package/jsep written in Typescript. tsep is not backwards compatible with jsep because jsep is a singleton whereas tsep uses parser engine instances so you can have many different parser configurations within one app.

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build tsep` to build the library.

## Running unit tests

Run `nx test tsep` to execute the unit tests via [Jest](https://jestjs.io).
