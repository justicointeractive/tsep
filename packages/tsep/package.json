{
  "name": "@tsep/tsep",
  "version": "0.0.1-alpha.2",
  "type": "commonjs"
}
