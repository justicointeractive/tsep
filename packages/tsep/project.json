{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/tsep/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/tsep",
        "main": "packages/tsep/src/index.ts",
        "tsConfig": "packages/tsep/tsconfig.lib.json",
        "assets": ["packages/tsep/*.md"]
      }
    },
    "publish": {
      "executor": "@nrwl/workspace:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs tsep {args.ver} {args.tag}"
      },
      "dependsOn": [
        {
          "projects": "self",
          "target": "build"
        }
      ]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/tsep/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/tsep"],
      "options": {
        "jestConfig": "packages/tsep/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
