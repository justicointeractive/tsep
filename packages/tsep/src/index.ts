export * from './lib/tsep';
export * from './lib/TsepError';
export * from './lib/types';
