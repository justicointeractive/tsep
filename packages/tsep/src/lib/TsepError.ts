export class TsepError extends Error {
  constructor(public description: string, public index: number) {
    super(description + ' at character ' + index);
  }
}
