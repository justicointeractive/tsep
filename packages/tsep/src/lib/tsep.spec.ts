import { Tsep } from './tsep';
import { TsepError } from './TsepError';
import { Expression } from './types';

describe('tsep', () => {
  it('should work', () => {
    expect(Tsep.create().parse(`foo(123)`)).toEqual({
      arguments: [{ raw: '123', type: 'Literal', value: 123 }],
      callee: { name: 'foo', type: 'Identifier' },
      type: 'CallExpression',
    });
  });
  it('should use separate operator lists', () => {
    expect(
      Tsep.create().addBinaryOp(':', 0, false).parse(`foo:bar(123)`)
    ).toEqual({
      left: { name: 'foo', type: 'Identifier' },
      operator: ':',
      right: {
        arguments: [{ raw: '123', type: 'Literal', value: 123 }],
        callee: { name: 'bar', type: 'Identifier' },
        type: 'CallExpression',
      },
      type: 'BinaryExpression',
    });
    expect(() => Tsep.create().parse(`foo:bar(123)`)).toThrowError(TsepError);
  });

  const defaultParser = Tsep.create();
  const testParser = (expr: string, match: DeepPartial<Expression>) =>
    expect(defaultParser.parse(expr)).toMatchObject(match);

  describe('Expression Parser', () => {
    it('Constants', function () {
      testParser("'abc'", { value: 'abc' });
      testParser('"abc"', { value: 'abc' });
      testParser('123', { value: 123 });
      testParser('12.3', { value: 12.3 });
    });
  });

  describe('Literal Parsing', () => {
    [
      ["'a \\w b'", 'a w b'],
      ["'a \\' b'", "a ' b"],
      ["'a \\n b'", 'a \n b'],
      ["'a \\r b'", 'a \r b'],
      ["'a \\t b'", 'a \t b'],
      ["'a \\b b'", 'a \b b'],
      ["'a \\f b'", 'a \f b'],
      ["'a \\v b'", 'a \v b'],
      ["'a \\ b'", 'a  b'],
    ].forEach((t) =>
      it(`Should parse ${t[0]}`, () => {
        testParser(t[0], { value: t[1], raw: t[0] });
      })
    );
  });

  it('Variables', function () {
    testParser('abc', { name: 'abc' });
    testParser('a.b[c[0]]', {
      property: {
        type: 'MemberExpression',
      },
    });
    testParser('Δέλτα', { name: 'Δέλτα' });
    testParser('a?.b?.(arg)?.[c] ?. d', {
      type: 'MemberExpression',
      optional: true,
    });
  });

  it('Function Calls', function () {
    testParser('a(b, c(d,e), f)', {});
    testParser('a b + c', {});
    testParser("'a'.toString()", {});
    testParser('[1].length', {});
    testParser(';', {});
    // allow all spaces or all commas to separate arguments
    testParser('check(a, b, c, d)', {});
    testParser('check(a b c d)', {});
  });

  it('Arrays', function () {
    testParser('[]', { type: 'ArrayExpression', elements: [] });

    testParser('[a]', {
      type: 'ArrayExpression',
      elements: [{ type: 'Identifier', name: 'a' }],
    });
  });

  describe('Ops', () => {
    const scopedParser = Tsep.create().addBinaryOp('**', 11, true);

    [
      '1',
      '1+2',
      '1*2',
      '1*(2+3)',
      '(1+2)*3',
      '(1+2)*3+4-2-5+2/2*3',
      '1 + 2-   3*	4 /8',
      '\n1\r\n+\n2\n',
      '1 + -2',
      '-1 + -2 * -3 * 2',
      '2 ** 3 ** 4',
      '2 ** 3 ** 4 * 5 ** 6 ** 7 * (8 + 9)',
      '(2 ** 3) ** 4 * (5 ** 6 ** 7) * (8 + 9)',
    ].forEach((expr) =>
      it(`Expr: ${expr.split('\r').join('\\r')}`, () =>
        expect(scopedParser.parse(expr)).toMatchSnapshot())
    );
  });

  it('Custom operators', () => {
    const scopedParser = Tsep.create();

    const testParser = (expr: string, match: DeepPartial<Expression>) =>
      expect(scopedParser.parse(expr)).toMatchObject(match);

    scopedParser.addBinaryOp('^', 10);
    testParser('a^b', {});
    scopedParser.removeBinaryOp('^');

    scopedParser.addBinaryOp('×', 9);
    testParser('a×b', {
      type: 'BinaryExpression',
      left: { name: 'a' },
      right: { name: 'b' },
    });
    scopedParser.removeBinaryOp('×');

    scopedParser.addBinaryOp('or', 1);
    testParser('oneWord ordering anotherWord', {
      type: 'Compound',
      body: [
        {
          type: 'Identifier',
          name: 'oneWord',
        },
        {
          type: 'Identifier',
          name: 'ordering',
        },
        {
          type: 'Identifier',
          name: 'anotherWord',
        },
      ],
    });
    scopedParser.removeBinaryOp('or');

    scopedParser.addUnaryOp('#');
    testParser('#a', {
      type: 'UnaryExpression',
      operator: '#',
      argument: { type: 'Identifier', name: 'a' },
    });
    scopedParser.removeUnaryOp('#');

    scopedParser.addUnaryOp('not');
    testParser('not a', {
      type: 'UnaryExpression',
      operator: 'not',
      argument: { type: 'Identifier', name: 'a' },
    });
    scopedParser.removeUnaryOp('not');

    scopedParser.addUnaryOp('notes');
    testParser('notes', {
      type: 'Identifier',
      name: 'notes',
    });
    scopedParser.removeUnaryOp('notes');
  });

  it('Custom alphanumeric operators', () => {
    const scopedParser = Tsep.create();

    const testParser = (expr: string, match: DeepPartial<Expression>) =>
      expect(scopedParser.parse(expr)).toMatchObject(match);

    scopedParser.addBinaryOp('and', 2);
    testParser('a and b', {
      type: 'BinaryExpression',
      operator: 'and',
      left: { type: 'Identifier', name: 'a' },
      right: { type: 'Identifier', name: 'b' },
    });
    testParser('bands', { type: 'Identifier', name: 'bands' });

    testParser('b ands', { type: 'Compound' });
    scopedParser.removeBinaryOp('and');

    scopedParser.addUnaryOp('not');
    testParser('not a', {
      type: 'UnaryExpression',
      operator: 'not',
      argument: { type: 'Identifier', name: 'a' },
    });
    testParser('notes', { type: 'Identifier', name: 'notes' });
    scopedParser.removeUnaryOp('not');
  });

  it('Custom identifier characters', () => {
    const scopedParser = Tsep.create();

    const testParser = (expr: string, match: DeepPartial<Expression>) =>
      expect(scopedParser.parse(expr)).toMatchObject(match);

    scopedParser.addIdentifierChar('@');
    testParser('@asd', {
      type: 'Identifier',
      name: '@asd',
    });
    scopedParser.removeIdentifierChar('@');
  });

  it('Bad Numbers', () => {
    testParser('1.', { type: 'Literal', value: 1, raw: '1.' });
    expect(() => {
      defaultParser.parse('1.2.3');
    }).toThrowError(TsepError);
  });

  function parseExprThrows(expr: string, name: string) {
    it(name, () => expectParseToThrow(expr));
  }

  function expectParseToThrow(expr: string) {
    expect(() => defaultParser.parse(expr)).toThrowError(TsepError);
  }

  describe('Missing arguments', () => {
    parseExprThrows('check(,)', 'detects missing argument (all)');
    parseExprThrows('check(,1,2)', 'detects missing argument (head)');
    parseExprThrows('check(1,,2)', 'detects missing argument (intervening)');
    parseExprThrows('check(1,2,)', 'detects missing argument (tail)');
    parseExprThrows('check(a, b c d) ', 'spaced arg after 1 comma');
    parseExprThrows('check(a, b, c d)', 'spaced arg at end');
    parseExprThrows('check(a b, c, d)', 'spaced arg first');
    parseExprThrows('check(a b c, d)', 'spaced args first');
  });

  describe('Uncompleted expression-call/array', () => {
    parseExprThrows('myFunction(a,b', 'detects unfinished expression call');
    parseExprThrows('[1,2', 'detects unfinished array');
    parseExprThrows('-1+2-', 'detects trailing operator');
  });

  [
    '!',
    '*x',
    '||x',
    '?a:b',
    '.',
    '()()',
    // '()', should throw 'unexpected )'...
    '() + 1',
  ].forEach((expr) => {
    parseExprThrows(expr, `should throw on invalid expr "${expr}"`);
  });

  describe('Hooks', () => {
    describe('gobble-spaces', () => {
      it('should allow manipulating what is considered whitespace', () => {
        const expr = 'a // skip all this';

        expectParseToThrow(expr);

        const parserInstance = Tsep.create();

        const testParser = (expr: string, match: DeepPartial<Expression>) =>
          expect(parserInstance.parse(expr)).toMatchObject(match);

        parserInstance.hooks.add('gobble-spaces', function () {
          if (this.char === '/' && this.expr.charAt(this.index + 1) === '/') {
            this.index += 2;
            while (!isNaN(this.code)) {
              this.index++;
            }
          }
        });
        testParser('a // skip all this', { type: 'Identifier' });
      });
    });

    describe('gobble-expression', () => {
      it('should accept this.node set by hook', () => {
        const expr = 'fn( 4 * 2';

        expectParseToThrow(expr);

        const parserInstance = Tsep.create();

        const testParser = (expr: string, match: DeepPartial<Expression>) =>
          expect(parserInstance.parse(expr)).toMatchObject(match);

        parserInstance.hooks.add(
          'gobble-expression',
          function (env) {
            if (this.char === 'f') {
              this.index += 9;
              env.node = { type: 'custom' };
            }
          },
          true
        );
        testParser(expr, { type: 'custom' });
      });

      it('should stop at first hook returning a node', () => {
        const expr = 'fn( 4 * 2';

        expectParseToThrow(expr);

        const parserInstance = Tsep.create();

        const testParser = (expr: string, match: DeepPartial<Expression>) =>
          expect(parserInstance.parse(expr)).toMatchObject(match);

        parserInstance.hooks.add(
          'gobble-expression',
          function (env) {
            if (this.char === 'f') {
              this.index += 9;
              env.node = { type: 'custom' };
            }
          },
          true
        );
        parserInstance.hooks.add('gobble-expression', function (env) {
          env.node = { type: 'wrong' };
        });
        testParser(expr, { type: 'custom' });
      });
    });

    describe('after-expression', () => {
      it('should allow altering an expression', () => {
        const parserInstance = Tsep.create();

        const testParser = (expr: string, match: DeepPartial<Expression>) =>
          expect(parserInstance.parse(expr)).toMatchObject(match);

        parserInstance.hooks.add('after-expression', function (env) {
          if (env.node) {
            env.node = { type: 'overruled' };
          }
        });
        testParser('1 + 2', { type: 'overruled' });
      });
    });

    describe('gobble-token', () => {
      it('should allow overriding gobbleToken', () => {
        const expr = '...';
        expectParseToThrow(expr);
        const parserInstance = Tsep.create();

        const testParser = (expr: string, match: DeepPartial<Expression>) =>
          expect(parserInstance.parse(expr)).toMatchObject(match);

        parserInstance.hooks.add('gobble-token', function (env) {
          if ([0, 1, 2].every((i) => this.expr.charAt(i) === '.')) {
            this.index += 3;
            env.node = { type: 'spread' };
          }
        });
        testParser(expr, { type: 'spread' });
      });

      it('should allow manipulating found token', () => {
        const after: (string | undefined)[] = [];
        const parserInstance = Tsep.create();

        parserInstance.hooks.add('after-token', function (env) {
          if (env.node) {
            env.node.type += ':)';
          }
          after.push(env.node?.type);
        });
        parserInstance.parse('a + 1 * !c(3) || d.e');
        expect(after.length).toEqual(4);
        expect(after[0]).toEqual('Identifier:)');
        expect(after[1]).toEqual('CallExpression:)');
        expect(after[2]).toEqual('UnaryExpression:)');
        expect(after[3]).toEqual('MemberExpression:)');
      });

      it('should stop processing hooks at first found node', () => {
        const expr = '...';

        expectParseToThrow(expr);

        const parserInstance = Tsep.create();

        const testParser = (expr: string, match: DeepPartial<Expression>) =>
          expect(parserInstance.parse(expr)).toMatchObject(match);

        parserInstance.hooks.add('gobble-token', function (env) {
          if ([0, 1, 2].every((i) => this.expr.charAt(i) === '.')) {
            this.index += 3;
            env.node = { type: 'spread' };
          }
        });
        parserInstance.hooks.add('gobble-token', function (env) {
          env.node = { type: 'wrong' };
        });
        testParser(expr, { type: 'spread' });
      });
    });
  });
});

type DeepPartial<T> = {
  [K in keyof T]?: DeepPartial<T[K]>;
};
