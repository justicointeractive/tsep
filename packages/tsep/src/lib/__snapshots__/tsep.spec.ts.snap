// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tsep Ops Expr: 
1\\r
+
2
 1`] = `
Object {
  "left": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "operator": "+",
  "right": Object {
    "raw": "2",
    "type": "Literal",
    "value": 2,
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: (1+2)*3 1`] = `
Object {
  "left": Object {
    "left": Object {
      "raw": "1",
      "type": "Literal",
      "value": 1,
    },
    "operator": "+",
    "right": Object {
      "raw": "2",
      "type": "Literal",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "operator": "*",
  "right": Object {
    "raw": "3",
    "type": "Literal",
    "value": 3,
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: (1+2)*3+4-2-5+2/2*3 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "left": Object {
          "left": Object {
            "left": Object {
              "raw": "1",
              "type": "Literal",
              "value": 1,
            },
            "operator": "+",
            "right": Object {
              "raw": "2",
              "type": "Literal",
              "value": 2,
            },
            "type": "BinaryExpression",
          },
          "operator": "*",
          "right": Object {
            "raw": "3",
            "type": "Literal",
            "value": 3,
          },
          "type": "BinaryExpression",
        },
        "operator": "+",
        "right": Object {
          "raw": "4",
          "type": "Literal",
          "value": 4,
        },
        "type": "BinaryExpression",
      },
      "operator": "-",
      "right": Object {
        "raw": "2",
        "type": "Literal",
        "value": 2,
      },
      "type": "BinaryExpression",
    },
    "operator": "-",
    "right": Object {
      "raw": "5",
      "type": "Literal",
      "value": 5,
    },
    "type": "BinaryExpression",
  },
  "operator": "+",
  "right": Object {
    "left": Object {
      "left": Object {
        "raw": "2",
        "type": "Literal",
        "value": 2,
      },
      "operator": "/",
      "right": Object {
        "raw": "2",
        "type": "Literal",
        "value": 2,
      },
      "type": "BinaryExpression",
    },
    "operator": "*",
    "right": Object {
      "raw": "3",
      "type": "Literal",
      "value": 3,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: (2 ** 3) ** 4 * (5 ** 6 ** 7) * (8 + 9) 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "left": Object {
          "raw": "2",
          "type": "Literal",
          "value": 2,
        },
        "operator": "**",
        "right": Object {
          "raw": "3",
          "type": "Literal",
          "value": 3,
        },
        "type": "BinaryExpression",
      },
      "operator": "**",
      "right": Object {
        "raw": "4",
        "type": "Literal",
        "value": 4,
      },
      "type": "BinaryExpression",
    },
    "operator": "*",
    "right": Object {
      "left": Object {
        "raw": "5",
        "type": "Literal",
        "value": 5,
      },
      "operator": "**",
      "right": Object {
        "left": Object {
          "raw": "6",
          "type": "Literal",
          "value": 6,
        },
        "operator": "**",
        "right": Object {
          "raw": "7",
          "type": "Literal",
          "value": 7,
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "operator": "*",
  "right": Object {
    "left": Object {
      "raw": "8",
      "type": "Literal",
      "value": 8,
    },
    "operator": "+",
    "right": Object {
      "raw": "9",
      "type": "Literal",
      "value": 9,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: -1 + -2 * -3 * 2 1`] = `
Object {
  "left": Object {
    "argument": Object {
      "raw": "1",
      "type": "Literal",
      "value": 1,
    },
    "operator": "-",
    "prefix": true,
    "type": "UnaryExpression",
  },
  "operator": "+",
  "right": Object {
    "left": Object {
      "left": Object {
        "argument": Object {
          "raw": "2",
          "type": "Literal",
          "value": 2,
        },
        "operator": "-",
        "prefix": true,
        "type": "UnaryExpression",
      },
      "operator": "*",
      "right": Object {
        "argument": Object {
          "raw": "3",
          "type": "Literal",
          "value": 3,
        },
        "operator": "-",
        "prefix": true,
        "type": "UnaryExpression",
      },
      "type": "BinaryExpression",
    },
    "operator": "*",
    "right": Object {
      "raw": "2",
      "type": "Literal",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: 1 + -2 1`] = `
Object {
  "left": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "operator": "+",
  "right": Object {
    "argument": Object {
      "raw": "2",
      "type": "Literal",
      "value": 2,
    },
    "operator": "-",
    "prefix": true,
    "type": "UnaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: 1 + 2-   3*	4 /8 1`] = `
Object {
  "left": Object {
    "left": Object {
      "raw": "1",
      "type": "Literal",
      "value": 1,
    },
    "operator": "+",
    "right": Object {
      "raw": "2",
      "type": "Literal",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "operator": "-",
  "right": Object {
    "left": Object {
      "left": Object {
        "raw": "3",
        "type": "Literal",
        "value": 3,
      },
      "operator": "*",
      "right": Object {
        "raw": "4",
        "type": "Literal",
        "value": 4,
      },
      "type": "BinaryExpression",
    },
    "operator": "/",
    "right": Object {
      "raw": "8",
      "type": "Literal",
      "value": 8,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: 1 1`] = `
Object {
  "raw": "1",
  "type": "Literal",
  "value": 1,
}
`;

exports[`tsep Ops Expr: 1*(2+3) 1`] = `
Object {
  "left": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "operator": "*",
  "right": Object {
    "left": Object {
      "raw": "2",
      "type": "Literal",
      "value": 2,
    },
    "operator": "+",
    "right": Object {
      "raw": "3",
      "type": "Literal",
      "value": 3,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: 1*2 1`] = `
Object {
  "left": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "operator": "*",
  "right": Object {
    "raw": "2",
    "type": "Literal",
    "value": 2,
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: 1+2 1`] = `
Object {
  "left": Object {
    "raw": "1",
    "type": "Literal",
    "value": 1,
  },
  "operator": "+",
  "right": Object {
    "raw": "2",
    "type": "Literal",
    "value": 2,
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: 2 ** 3 ** 4 * 5 ** 6 ** 7 * (8 + 9) 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "raw": "2",
        "type": "Literal",
        "value": 2,
      },
      "operator": "**",
      "right": Object {
        "left": Object {
          "raw": "3",
          "type": "Literal",
          "value": 3,
        },
        "operator": "**",
        "right": Object {
          "raw": "4",
          "type": "Literal",
          "value": 4,
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
    "operator": "*",
    "right": Object {
      "left": Object {
        "raw": "5",
        "type": "Literal",
        "value": 5,
      },
      "operator": "**",
      "right": Object {
        "left": Object {
          "raw": "6",
          "type": "Literal",
          "value": 6,
        },
        "operator": "**",
        "right": Object {
          "raw": "7",
          "type": "Literal",
          "value": 7,
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "operator": "*",
  "right": Object {
    "left": Object {
      "raw": "8",
      "type": "Literal",
      "value": 8,
    },
    "operator": "+",
    "right": Object {
      "raw": "9",
      "type": "Literal",
      "value": 9,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`tsep Ops Expr: 2 ** 3 ** 4 1`] = `
Object {
  "left": Object {
    "raw": "2",
    "type": "Literal",
    "value": 2,
  },
  "operator": "**",
  "right": Object {
    "left": Object {
      "raw": "3",
      "type": "Literal",
      "value": 3,
    },
    "operator": "**",
    "right": Object {
      "raw": "4",
      "type": "Literal",
      "value": 4,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;
